src/testModel.ts

```diff
@@ -91,24 +93,24 @@ export class TestModel extends DisposableBase {
     this._useLegacyCLIDriver = playwrightInfo.version < 1.44;
     this._playwrightTest =  this._useLegacyCLIDriver ? new PlaywrightTestCLI(this._vscode, this, collection.embedder) : new PlaywrightTestServer(this._vscode, this, collection.embedder);
     this.tag = new this._vscode.TestTag(this.config.configFile);
-    this._spawnTraceViewer = new SpawnTraceViewer(this._vscode, this._embedder.envProvider, this.config);
+    this._traceViewers = [
+      new EmbeddedTraceViewer(this._vscode, this._embedder.settingsModel, this._embedder.context.extensionUri, this.config, this._playwrightTest),
+      new SpawnTraceViewer(this._vscode, this._embedder.settingsModel, this._embedder.envProvider, this.config),
+    ];
 
     this._disposables = [
       this._embedder.settingsModel.showTrace.onChange(() => this._closeTraceViewerIfNeeded()),
+      this._embedder.settingsModel.embeddedTraceViewer.onChange(() => this._closeTraceViewerIfNeeded()),
       this._collection.onUpdated(() => this._closeTraceViewerIfNeeded()),
     ];
   }
 
   traceViewer() {
-    if (!this._embedder.settingsModel.showTrace.get())
-      return;
-    if (this._spawnTraceViewer.checkVersion())
-      return this._spawnTraceViewer;
-  }
+    if (this._currentTraceViewer !== undefined)
+      return this._currentTraceViewer;
 
-  _closeTraceViewerIfNeeded() {
-    if (this._collection.selectedModel() !== this || !this._embedder.settingsModel.showTrace.get())
-      this._spawnTraceViewer.close();
+    this._currentTraceViewer = this._traceViewers.find(t => t.checkSupport(true)) ?? null;
```

Comment:
It does not look like this is necessarily a user gesture. Looks like a generic programmatic getter to me

Reply:
Addressed in 31a15f3f926c5f17a8e16a4476d52c62f11778a9 with `traceViewer()` and `ensureTraceViewer()`. The latter creates a trace viewer if not initialized, as a result of a user gesture (running tests or selecting a test item).